sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java

```diff
@@ -211,56 +213,64 @@ Mono<StoreResponse> writePrivateAsync(
-                                                   }
-                                               }
-                                           );
+                return Mono.defer(() -> this.transportClient.invokeResourceOperationAsync(primaryUri, request)
+                        .doOnError(
+                            t -> {
+                                try {
+                                    Throwable unwrappedException = Exceptions.unwrap(t);
+                                    CosmosException ex = Utils.as(unwrappedException, CosmosException.class);
+                                    Exception rawException = null;
+                                    if (ex == null) {
+                                        rawException = Utils.as(unwrappedException, Exception.class);
+
+                                        if (rawException == null) {
+                                            throw unwrappedException;
+                                        }
+                                    }
+
+                                    storeReader.createAndRecordStoreResult(
+                                        request,
+                                        null, ex != null ? ex : rawException,
+                                        false,
+                                        false,
+                                        primaryUri,
+                                        replicaStatusList);
+                                    String value = ex != null ?
+                                        ex.getResponseHeaders().get(HttpConstants.HttpHeaders.WRITE_REQUEST_TRIGGER_ADDRESS_REFRESH) : null;
+                                        if (!Strings.isNullOrWhiteSpace(value)) {
+                                            Integer result = Integers.tryParse(value);
+                                            if (result != null && result == 1) {
+                                                startBackgroundAddressRefresh(request);
+                                            }
+                                        }
+                                    } catch (Throwable throwable) {
+                                        if (throwable instanceof Error) {
+                                            logger.error("Unexpected failure in handling orig [{}]", t.getMessage(), t);
+                                            logger.error("Unexpected failure in handling orig [{}] : new [{}]", t.getMessage(), throwable.getMessage(), throwable);
+                                            throw (Error) throwable;
+                                        } else {
+                                            // this happens before any retry policy - like for example GoneAndRetryRetryPolicy
+                                            // kicks in - no need to spam warn/error level logs yet
+                                            logger.info("Unexpected failure in handling orig [{}]", t.getMessage(), t);
```

Comment:
Why both - isn't this guaranteed to always be a duplicate?

Reply:
Fixed in next iteration.